modules/dhcp/providers/server/isc.rb

```diff
@@ -104,38 +96,72 @@ def loadSubnetData subnet
-          # delete expired (explicitly) expired leases
-          if opts[:state] == "free" || (opts[:next_state] == "free" && opts[:ends] && opts[:ends] < Time.now)
-            record = find_record_by_ip(subnet, ip, 'lease')
-            subnet.delete record if record
+
+          subnet = @service.find_subnet(ip)
+          next unless subnet
+          ret_val << Proxy::DHCP::Lease.new(opts.merge(:subnet => subnet, :ip => ip))
+        end
+      end
+
+      ret_val
+    end
+
+    def initialize_memory_store_with_dhcp_records(records)
+      records.each do |record|
+
+        case record
+        when Proxy::DHCP::Reservation
+          if record.options[:deleted]
+            record = @service.find_host_by_hostname(record.subnet_address, record.name)
+            @service.delete_host(record) if record
             next
           end
-          if subnet.include? ip
-            # delete possible duplicities 
-            if dupe = subnet.has_mac?(opts[:mac], :lease)
-              subnet.delete(dupe)
-            end
-            if dupe = subnet.has_ip?(ip, :lease)
-              subnet.delete(dupe)
-            end
-            # and add the record
-            Proxy::DHCP::Lease.new(opts.merge(:subnet => subnet, :ip => ip))
+          if dupe = @service.find_host_by_mac(record.subnet_address, record.mac)
+            @service.delete_host(dupe)
           end
+          if dupe = @service.find_host_by_ip(record.subnet_address, record.ip)
+            @service.delete_host(dupe)
+          end
+
+          @service.add_host(record.subnet_address, record)
+        when Proxy::DHCP::Lease
+          if record.options[:state] == "free" || (record.options[:next_state] == "free" && record.options[:ends] && record.options[:ends] < Time.now)
+            record = @service.find_lease_by_ip(record.subnet_address, record.ip)
+            @service.delete_lease(record) if record
+            next
+          end
+
+          if dupe = @service.find_lease_by_mac(record.subnet_address, record.mac)
+            @service.delete_lease(dupe)
+          end
+          if dupe = @service.find_lease_by_ip(record.subnet_address, record.ip)
+            @service.delete_lease(dupe)
+          end
+
+          @service.add_lease(record.subnet_address, record)
         end
       end
-      report "Enumerated hosts on #{subnet.network}"
     end
 
-    private
+    def loadSubnetData subnet
+      initialize_memory_store_with_dhcp_records(parse_config_and_leases_for_records)
+    end
+
+    def parse_config_for_subnets
+      ret_val = []
 
-    def loadSubnets
-      super
       @config.scan(/subnet\s+([\d\.]+)\s+netmask\s+([\d\.]+)/) do |match|
         next unless managed_subnet? "#{match[0]}/#{match[1]}"
 
-        Proxy::DHCP::Subnet.new(self, match[0], match[1])
+        ret_val << Proxy::DHCP::Subnet.new(match[0], match[1])
       end
-      "Enumerated the scopes on #{@name}"
+
+      ret_val
+    end
+
+    def loadSubnets
```

Comment:
It's a legacy name, I'd like to minimize renames in older parts of the code at this point.